vlrouter.go

```diff
@@ -0,0 +1,662 @@
+
+	// Set Mac addresses
+	ipFlow.SetMacDa(destMacAddr)
+	ipFlow.SetMacSa(self.myRouterMac)
+
+	// Point the route at output port
+	err = ipFlow.Next(outPort)
+	if err != nil {
+		log.Errorf("Error installing flow for endpoint: %+v. Err: %v", endpoint, err)
+		return err
+	}
+
+	// Store the flow
+	self.flowDb[endpoint.IpAddr.String()] = ipFlow
+
+	if endpoint.EndpointType == "internal-bgp" {
+		return nil
+	}
+
+	//dial grpc server
+	conn, err := grpc.Dial("127.0.0.1:8080", grpc.WithBlock(), grpc.WithInsecure())
+	if err != nil {
+		log.Fatalf("did not connect: %v", err)
+	}
+	defer conn.Close()
+
+	path := &api.Path{
+		Pattrs: make([][]byte, 0),
+	}
+
+	// form the path structure with appropriate path attributes
+	nlri := bgp.NewIPAddrPrefix(32, endpoint.IpAddr.String())
+	path.Nlri, _ = nlri.Serialize()
+	origin, _ := bgp.NewPathAttributeOrigin(bgp.BGP_ORIGIN_ATTR_TYPE_EGP).Serialize()
+	path.Pattrs = append(path.Pattrs, origin)
+	aspathParam := []bgp.AsPathParamInterface{bgp.NewAs4PathParam(2, []uint32{65002})}
+	aspath, _ := bgp.NewPathAttributeAsPath(aspathParam).Serialize()
+	path.Pattrs = append(path.Pattrs, aspath)
+	n, _ := bgp.NewPathAttributeNextHop(self.agent.routerIP).Serialize()
+	path.Pattrs = append(path.Pattrs, n)
+
+	name := ""
+
+	arg := &api.ModPathArguments{
+		Resource: api.Resource_GLOBAL,
+		Name:     name,
+		Paths:    []*api.Path{path},
+	}
+
+	//send arguement stream
+	client := api.NewGobgpApiClient(conn)
+	if client == nil {
+		log.Errorf("Gobgpapi stream invalid")
+		return nil
+	}
+
+	stream, err := client.ModPath(context.Background())
+	if err != nil {
+		log.Errorf("Fail to enforce Modpathi", err)
+		return err
+	}
+
+	err = stream.Send(arg)
+	if err != nil {
+		log.Errorf("Failed to send strean", err)
+		return err
+	}
+	stream.CloseSend()
+	res, e := stream.CloseAndRecv()
+	if e != nil {
+		log.Errorf("Falied toclose stream ")
+		return e
+	}
+	if res.Code != api.Error_SUCCESS {
+		return fmt.Errorf("error: code: %d, msg: %s", res.Code, res.Msg)
+	}
+
+	return nil
+}
+
+/* RemoveLocalEndpoint does the following
+1) Removes the local endpoint and associated flows from OVS
+2) Withdraws the route from BGP RIB
+*/
+func (self *Vlrouter) RemoveLocalEndpoint(endpoint OfnetEndpoint) error {
+
+	// Remove the port vlan flow.
+	portVlanFlow := self.portVlanFlowDb[endpoint.PortNo]
+	if portVlanFlow != nil {
+		err := portVlanFlow.Delete()
+		if err != nil {
+			log.Errorf("Error deleting portvlan flow. Err: %v", err)
+		}
+	}
+
+	// Find the flow entry
+	ipFlow := self.flowDb[endpoint.IpAddr.String()]
+	if ipFlow == nil {
+		log.Errorf("Error finding the flow for endpoint: %+v", endpoint)
+		return errors.New("Flow not found")
+	}
+
+	// Delete the Fgraph entry
+	err := ipFlow.Delete()
+	if err != nil {
+		log.Errorf("Error deleting the endpoint: %+v. Err: %v", endpoint, err)
+	}
+
+	//dial grpc server
+	conn, err := grpc.Dial("127.0.0.1:8080", grpc.WithBlock(), grpc.WithInsecure())
```

Comment:
Again, it would be good to move this below part into a function and define a clear API between vlrouter and BGP module